pipelines/pipeline.yml

```diff
@@ -186,35 +134,157 @@ stages:
+                      echo "${lease_id}" > ./lease_id.txt
+                    env:
+                      NAMESPACE=$(namespace)
+
+                # Install Terraform
+                - task: TerraformInstaller@0
+                  inputs:
+                    terraformVersion: "0.12.18"
+                  displayName: "Install Terraform"
+
+                # Configure the Terraform backend
+                - bash: ./scripts/create-tf-backend.sh
+                  displayName: "Configure Terraform Backend"
+
+                # terraform init
+                - bash: |
+                    cd modules
+                    cat ./backend.tf
+                    terraform init -input=false
+                    terraform apply \
+                      -auto-approve \
+                      -input=false \
+                      -var="namespace=${NAMESPACE}" \
+                      -var="budget_notification_from_email=${NOTIFY_EMAIL}"
+                  displayName: "Terraform Init/Apply"
+                  env:
+                    NAMESPACE: $(namespace)
+                    NOTIFY_EMAIL: $(NOTIFY_EMAIL)
+
+                # Deploy Application Code to AWS
+                - bash: |
+                    ARTIFACTS_BUCKET_NAME=$(cd modules && terraform output artifacts_bucket_name)
+                    NAMESPACE=$(cd modules && terraform output namespace)
+                    ./scripts/deploy.sh \
+                      $(Pipeline.Workspace)/bin/build_artifacts.zip \
+                      ${NAMESPACE} \
+                      ${ARTIFACTS_BUCKET_NAME}
+                  displayName: "Deploy Application Code"
+                  env:
+                    AWS_DEFAULT_REGION: us-east-1
+
+
+                # Functional Tests
+                - bash: |
+                    set -euxo pipefail
+                    mkdir -p junit-report
+                    # Run functional tests
+                    go test -v ./tests/... 2>&1 | tee >(go-junit-report > junit-report/functional.xml)
+                  displayName: "Functional Tests"
+
 
                 # Publish junit test results (for unit and functional tests)
                 - task: PublishTestResults@2
                   inputs:
                     testRunner: JUnit
-                    testResultsFiles: $(Pipeline.Workspace)/junit-report/*.xml
+                    testResultsFiles: $(System.DefaultWorkingDirectory)/junit-report/*.xml
                     testRunTitle: Functional Tests
                   condition: always()
+
+                # Publish the dce-cli executable, so we can use
+                # it in our Cleanup stage
+                - publish: $(System.DefaultWorkingDirectory)/dce
+                  artifact: dce-cli
+                - publish: $(System.DefaultWorkingDirectory)/dce.yml
+                  artifact: dce-yml
+                - publish: $(System.DefaultWorkingDirectory)/lease_id.txt
```

Comment:
Can't we just requery this?

Reply:
yeah, we could. There are just some things in the current version of dce-cli that make it difficult to lookup a specific lease (printing to stderr, can't query by status and principal..), so I was trying to avoid it for now.